### ISSUE TITLE
Postgres `timetz` always stored in UTC

### ISSUE THREAD

**Issue by giraffesyo** [prisma/prisma#7915]
### Bug description

Currently, it seems any time you create a record with type `timetz` the time is automatically converted to be in UTC. In my case, I would like all of the database records to be stored in `America/Chicago` time zone, so I'm preparing the ISO string in that way before using the `create` method. However, prisma doesn't respect this and converts the time back to UTC. 

The create statement is below:

```typescript
   console.log(timeEST) // 2021-06-28T12:00:00.000-05:00
    const course = await prisma.classes.create({
      data: {
        day_of_week,
        class_time: timeEST,
        instructors: { connect: { id: instructor } },
        class_types: { connect: { id: class_type } },
      },
    })
```
Database results all showing +0 zone: 
```devidgmefdn94u=> select * from classes;
 id | day_of_week | class_time  | instructor | class_type | virtual 
----+-------------+-------------+------------+------------+---------
 19 |           5 | 11:30:00+00 |          3 |          7 | t
 20 |           5 | 11:30:00+00 |          3 |          7 | t
 35 |           1 | 17:00:00+00 |          3 |          7 | t
 36 |           1 | 17:00:00+00 |          3 |          7 | t
 37 |           1 | 17:00:00+00 |          3 |          7 | t
 38 |           1 | 17:00:00+00 |          3 |          7 | t
(6 rows)
```

Model for this:

```prisma
model classes {
  id            Int             @id @default(autoincrement())
  day_of_week   Int
  class_time    DateTime        @db.Timetz(6)
  instructor    Int
  class_type    Int
  virtual       Boolean?        @default(true)
  class_types   class_types     @relation(fields: [class_type], references: [id])
  instructors   instructors     @relation(fields: [instructor], references: [id])
  zoom_meetings zoom_meetings[]
}
```

### Environment & setup

- OS: Mac OS
- Database: Postgres 13.3
- Node.js version: v12.22.1

### Prisma Version

```
prisma               : 2.25.0
@prisma/client       : 2.25.0
Current platform     : darwin
Query Engine         : query-engine c838e79f39885bc8e1611849b1eb28b5bb5bc922 (at node_modules/@prisma/engines/query-engine-darwin)
Migration Engine     : migration-engine-cli c838e79f39885bc8e1611849b1eb28b5bb5bc922 (at node_modules/@prisma/engines/migration-engine-darwin)
Introspection Engine : introspection-core c838e79f39885bc8e1611849b1eb28b5bb5bc922 (at node_modules/@prisma/engines/introspection-engine-darwin)
Format Binary        : prisma-fmt c838e79f39885bc8e1611849b1eb28b5bb5bc922 (at node_modules/@prisma/engines/prisma-fmt-darwin)
Default Engines Hash : c838e79f39885bc8e1611849b1eb28b5bb5bc922
Studio               : 0.402.0
```

**Comment by kladnik** [comment]
I am having the same issues. Opened an issue here: #7825 
This is definitely a bug.

**Comment by ferjul17** [comment]
Hi,

I think this is not a Prisma issue, but a PostgreSQL one.

As explained in their [documentation](https://www.postgresql.org/docs/current/datatype-datetime.html#id-1.5.7.13.18.7.4):
> For timestamp with time zone, the internally stored value is always in UTC (Universal Coordinated Time, traditionally known as Greenwich Mean Time, GMT). An input value that has an explicit time zone specified is converted to UTC using the appropriate offset for that time zone. If no time zone is stated in the input string, then it is assumed to be in the time zone indicated by the system's TimeZone parameter, and is converted to UTC using the offset for the timezone zone.

PostgreSQL does not store the timezone, and I do not expect Prisma to do it for me.

I think the issue can be closed.

**Comment by kladnik** [comment]
The timestamps are always saved in UTC that is not the issue, it's how it's handled when reading them. 
See the examples I posted in the related issue.

**Comment by ferjul17** [comment]
I agree with you, your issue is really an issue of Prisma. But I don't think this one it is.

**Comment by kladnik** [comment]
> I agree with you, your issue is really an issue of Prisma. But I don't think this one it is.

You might be actually correct, since time doesn't exists without tz. I do wonder why they closed my issue claiming it's a duplicate of this

**Comment by giraffesyo** [comment]
> Hi,
> 
> 
> 
> I think this is not a Prisma issue, but a PostgreSQL one.
> 
> 
> 
> As explained in their [documentation](https://www.postgresql.org/docs/current/datatype-datetime.html#id-1.5.7.13.18.7.4):
> 
> > For timestamp with time zone, the internally stored value is always in UTC (Universal Coordinated Time, traditionally known as Greenwich Mean Time, GMT). An input value that has an explicit time zone specified is converted to UTC using the appropriate offset for that time zone. If no time zone is stated in the input string, then it is assumed to be in the time zone indicated by the system's TimeZone parameter, and is converted to UTC using the offset for the timezone zone.
> 
> 
> 
> PostgreSQL does not store the timezone, and I do not expect Prisma to do it for me.
> 
> 
> 
> I think the issue can be closed.

Postgres supports storing a time zone. The way postgres internally stores the time string isn't important because it also stores the time zone specified. When you query the record from the database it includes the associated time zone information in the result but Prisma drops it.

**Comment by aqrln** [comment]
Related: https://github.com/prisma/prisma/issues/7917

**Comment by SevInf** [comment]
Internal note: https://www.notion.so/prismaio/Timezone-issues-in-query-engine-and-client-a1a2a9c6096a49f9a63db50c4c68bfe6

**Comment by diakonovm** [comment]
Any solution to this issue?

**Comment by rstrand** [comment]
> > Hi,
> > I think this is not a Prisma issue, but a PostgreSQL one.
> > As explained in their [documentation](https://www.postgresql.org/docs/current/datatype-datetime.html#id-1.5.7.13.18.7.4):
> > > For timestamp with time zone, the internally stored value is always in UTC (Universal Coordinated Time, traditionally known as Greenwich Mean Time, GMT). An input value that has an explicit time zone specified is converted to UTC using the appropriate offset for that time zone. If no time zone is stated in the input string, then it is assumed to be in the time zone indicated by the system's TimeZone parameter, and is converted to UTC using the offset for the timezone zone.
> > 
> > 
> > PostgreSQL does not store the timezone, and I do not expect Prisma to do it for me.
> > I think the issue can be closed.
> 
> Postgres supports storing a time zone. The way postgres internally stores the time string isn't important because it also stores the time zone specified. When you query the record from the database it includes the associated time zone information in the result but Prisma drops it.

With a timestamp with timezone PostgreSQL always returns the time in the zone the database is configured to use and not the timezone that the timestamp was saved with. Postgres doesn't store the zone the timestamp was saved with and stores it internally as UTC.

**Comment by mihailShumilov** [comment]
Any updates/patches for this?

**Comment by mihailShumilov** [comment]
I have prepared fix for that

please take a look at 
https://github.com/prisma/prisma/pull/28071

**Comment by GeetanshVirtuos** [comment]
Hey! Has this been fixed yet?

**Comment by UgaTheDev** [comment]
Still reproduces on `v7` (`05ce49014`). Root cause in the 7.x TypeScript adapter:
`packages/adapter-pg/src/conversion.ts:426` routes `TIMETZ` to `formatTime` (`:470`),
which emits `HH:MM:SS` with **no offset suffix**. PostgreSQL then resolves that bare
time using the session `TimeZone`, so a value the adapter computed in UTC gets stored
as local wall time. With the session at `Asia/Kolkata`, `new Date('1970-01-01T10:30:00Z')`
is stored as `10:30:00+05:30` — i.e. `05:00:00` UTC, a 5h30m shift.

Same in `adapter-neon/src/conversion.ts:412`. `adapter-ppg` is not affected on write:
it sends `toISOString()` (`:55-56`), which carries an explicit `Z`.

One thing worth flagging for whoever picks this up: this bug and #7917 are mirror
images, so a same-session write→read round-trip **passes** — the write drops the offset
and the read strips it, and the string returns unchanged. What is actually wrong is the
value stored in the database and every read of a value written by any other client. Any
regression test needs to assert on `"col"::text` with a non-UTC session `TimeZone`,
not just on the round-tripped value.

There is also a semantics question the team should settle before a fix, because the two
issues want different things: this issue asks for the caller's offset to be **preserved**
as-is, whereas Prisma models `Timetz` as `DateTime` (a UTC instant), which implies
normalising to `+00:00`. Writing an explicit `+00:00` is the minimal, non-breaking fix
and makes the stored instant correct; preserving an arbitrary caller offset would need a
representation that survives the `DateTime` mapping. I'd like to implement the former.

Note #28071 was closed unmerged and only touched `timestamptz`, not `timetz`.

The sibling `timestamptz` fix is open in #29481 (same file, same class of bug — I
reviewed it there); happy to do this one.
